torch/csrc/jit/tensorexpr/codegen.cpp

```diff
@@ -83,21 +83,136 @@ void CodeGen::call_with_numel(void** args, int64_t numel) {
+// assign it an existing non-occupied memory allocation. If there are no such
+// allocations available, we'll create memory for it. Once we are beyond the
+// liveness range of this buffer, we'll mark its corresponding memory allocation
+// as "up for grabs" for future reuse.
+std::vector<std::pair<BufPtr, BufPtr>> linearScan(
+    const std::unordered_set<BufPtr>& bufs,
+    const std::unordered_map<BufPtr, std::tuple<int32_t, int32_t>>&
+        buf_ranges) {
+  // Sort buffers by the time they appear.
+  std::vector<BufPtr> bufs_sorted(bufs.begin(), bufs.end());
+  auto sorting_function_by_start_time = [&buf_ranges](
+                                            BufPtr b1, BufPtr b2) -> bool {
+    return std::get<0>(buf_ranges.at(b1)) < std::get<0>(buf_ranges.at(b2));
+  };
+  std::sort(
+      bufs_sorted.begin(), bufs_sorted.end(), sorting_function_by_start_time);
+
+  // Map intermediate buffers to the most recently used memory if any.
+  std::list<BufPtr> mem_up_for_grabs;
+  std::unordered_map<BufPtr, BufPtr> buf_mem_map;
+  std::vector<std::pair<BufPtr, BufPtr>> buf_allocs;
+
+  auto sorting_function_by_end_time = [&buf_ranges](
+                                          BufPtr b1, BufPtr b2) -> bool {
+    return std::get<1>(buf_ranges.at(b1)) < std::get<1>(buf_ranges.at(b2));
+  };
+  std::vector<BufPtr> buf_to_release;
+  for (auto buf : bufs_sorted) {
+    auto start = std::get<0>(buf_ranges.at(buf));
+    auto end = std::get<1>(buf_ranges.at(buf));
+
+    // Release memory for buffers whose liveness range ends before the creation
+    // time of this buf.
+    // TODO: optimize in-place opererations and copy operations
+    buf_to_release.clear();
+    for (auto& mapped : buf_mem_map) {
+      auto buf_mapped = mapped.first;
+      auto end_buf_mapped = std::get<1>(buf_ranges.at(buf_mapped));
+      if (end_buf_mapped < start) {
+        buf_to_release.push_back(buf_mapped);
+      }
+    }
+
+    // Sort the buffers in the order of used time so the head of the release
+    // list contains the most recently used buf.
+    std::sort(
+        buf_to_release.begin(),
+        buf_to_release.end(),
+        sorting_function_by_end_time);
+    for (auto& buf_rl : buf_to_release) {
+      mem_up_for_grabs.push_front(buf_mem_map[buf_rl]);
+      buf_mem_map.erase(buf_rl);
+    }
+
+    // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
+    // it, and there are no future reuses on its memory).
+    // TODO: reuse memory for bufs with dynamic shapes
+    if (!bufSize(buf)) {
+      buf_allocs.emplace_back(std::make_pair(buf, buf));
+      continue;
+    }
+
+    bool allocated = false;
+    // Check whether there are free memories that this buf can reuse.
+    for (auto it = mem_up_for_grabs.begin(); it != mem_up_for_grabs.end();
+         it++) {
+      auto m = *it;
+      if (bufSize(m) >= bufSize(buf)) {
+        buf_mem_map[buf] = m;
```

Comment:
Say we reused buf 'A' for buf 'B'. How do we mark 'A' from being reused again later? Could you please add a test to check that it works correctly?

I.e. if input IR is like this:
```
A = alloc(10);
B = alloc(10);
C = alloc(10);
first_use(A);
last_use(A)
first_use(B)
first_use(C)
last_use(C)
last_use(B)
Free(A)
Free(B)
Free(C)
```
We need to make sure we do not assign both B and C to A:
```
A = alloc(10);
B = alias(A);
C = alloc(10);
first_use(A);
last_use(A)
first_use(B)
first_use(C)
last_use(C)
last_use(B)
Free(A)
Free(C)
```

Reply:
All buffers that can be reused at the moment are in the list of `mem_up_for_grabs`. Once buf `A` is reused for buf `B`, buf `A` will be removed from `mem_up_for_grabs`. When we arrive the end of buf `B`'s  liveness range, buf `A` will be put back to `mem_up_for_grabs` for future reuse.

I'll add a test for this.

Reply:
> How does this comparison work if bufSize(m) returns c10::nullopt? Should we exclude such bufs from mem_up_for_grabs?

We will not put buffers with c10::nullopt bufSize to mem_up_for_grabs. Currently, we only reuse buffers with static shapes. For buffers with dynamic shapes, we will mark it as `allocate` in the following code:
```
     // If the buf has dynamic shapes, we'll skip it (i.e., allocate memory for
     // it, and there are no future reuses on its memory).
     // TODO: reuse memory for bufs with dynamic shapes
     if (!bufSize(buf)) {                                                                                                                                                      
       buf_allocs.emplace_back(std::make_pair(buf, buf));
       continue;
     }
```